proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,250 @@
+int, float, float4, etc. If the variable type is among any numerics or a type that
+cannot be an intangible type, there is only one exception where the register keyword
+can legally be applied to the variable. The only valid register type for such a 
+numeric variable type is 'c'. The register keyword in this statement:
+`float f : register(c0)`
+isn't binding a resource, rather it is specifying a constant register binding offset
+within the $Globals cbuffer, which is legacy behavior from DX9. Because this specific 
+register overload is only applicable if the developer wants to make modifications to the
+$Globals buffer, it is not useful when the overload is not used inside the $Globals
+context. That is, if this register overload appears within a `cbuffer {...}` or 
+`tbuffer {...}` scope, the register number will be ignored. In this case, a warning will
+be emitted that is treated as an error:
+`warning: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
+All other register types applied to such a type will emit an error diagnostic that 
+`"'%0' is an invalid resource type for register type '%1'"`, except for 'b' or 'i'.
+If the register type is 'b' or 'i', a warning will be emitted instead, that will be 
+treated as an error by default. <br>
+Below are some examples:
+
+| Code | Diagnostic |
+|-|-|
+| `float f : register(t0)` | "error: 'float' is an invalid resource type for register type 't'" |
+| `float f : register(c0)` | Valid, no errors
+| `float f : register(b0)` | "warning: deprecated legacy bool constant register binding 'b' used. 'b' is only used for constant buffer resource binding. Disable with --Wno-disallow-legacy-binding-rules" |
+| `float f : register(i0)` | "warning: deprecated legacy int constant register binding 'i' used. Disable with --Wno-disallow-legacy-binding-rules" |
+| `float f : register(x0)` | "error: register binding type 'x' not supported for variable of type 'float'" |
+| `cbuffer g_cbuffer { float f : register(c2); }` | "warning: register binding 'c' should only be used in global contexts. Disable with --Wno-disallow-legacy-binding-rules" |
+| `struct Foo { RWBuffer<float> f; }; Foo foo : register(c2);`| "warning: 'c' register type should only be used for basic types. Disable with --Wno-disallow-legacy-binding-rules" |
+| `RWBuffer<float> f : register(c3);`| "error: invalid register specification, expected 'u' binding" |
+
+
+## Detailed design
+
+All the compiler has to work with is a Decl object and the context the decl appears in.
+From this information, the first goal is to set a specific set of flags that can fully inform
+the right diagnostic to emit. The first group of flags are the decl type flags, which
+are either `basic`, `resource`, `udt`, or `other`. `basic` refers to a numeric variable
+type. `other` refers to a type that cannot possibly be resource, because it lacks the 
+resource attribute.<br>

```

Comment:
`other` actually refers to other special object types (intangible types) that are not resources, and thus cannot be placed into a cbuffer or bound as a resource.